x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec

```diff
@@ -0,0 +1,594 @@
+sample_data_str     | 2023-10-23T13:51:54.732Z  |  172.21.3.15      |  725448               |  Connection error
+sample_data_str     | 2023-10-23T13:33:34.937Z  |  172.21.0.5       |  1232382              |  Disconnected
+sample_data_str     | 2023-10-23T12:27:28.948Z  |  172.21.2.113     |  2764889              |  Connected to 10.1.0.2
+sample_data_str     | 2023-10-23T12:15:03.360Z  |  172.21.2.162     |  3450233              |  Connected to 10.1.0.3
+sample_data_ts_long | 2023-10-23T13:55:01.543Z  |  172.21.3.15      |  1756467              |  Connected to 10.1.0.1
+sample_data_ts_long | 2023-10-23T13:53:55.832Z  |  172.21.3.15      |  5033755              |  Connection error
+sample_data_ts_long | 2023-10-23T13:52:55.015Z  |  172.21.3.15      |  8268153              |  Connection error
+sample_data_ts_long | 2023-10-23T13:51:54.732Z  |  172.21.3.15      |  725448               |  Connection error
+sample_data_ts_long | 2023-10-23T13:33:34.937Z  |  172.21.0.5       |  1232382              |  Disconnected
+sample_data_ts_long | 2023-10-23T12:27:28.948Z  |  172.21.2.113     |  2764889              |  Connected to 10.1.0.2
+sample_data_ts_long | 2023-10-23T12:15:03.360Z  |  172.21.2.162     |  3450233              |  Connected to 10.1.0.3
+;
+
+multiIndexWhereIpStringTsLong
+required_capability: union_types
+required_capability: metadata_fields
+
+FROM sample_data* METADATA _index
+| WHERE TO_LONG(@timestamp) < 1698068014937 AND TO_STRING(client_ip) == "172.21.2.162"
+| KEEP _index, event_duration, message
+| SORT _index ASC, message ASC
+;
+
+_index:keyword       |  event_duration:long  |  message:keyword
+sample_data          |  3450233              |  Connected to 10.1.0.3
+sample_data_str      |  3450233              |  Connected to 10.1.0.3
+sample_data_ts_long  |  3450233              |  Connected to 10.1.0.3
+;
+
+multiIndexWhereIpStringTsLongStats
+required_capability: union_types
+
+FROM sample_data*
+| WHERE TO_LONG(@timestamp) < 1698068014937 AND TO_STRING(client_ip) == "172.21.2.162"
+| STATS count=count(*) BY message
+| SORT count DESC, message ASC
+| KEEP count, message
+;
+
+count:long  |  message:keyword
+3           |  Connected to 10.1.0.3
+;
+
+multiIndexWhereIpStringLikeTsLong
+required_capability: union_types
+required_capability: metadata_fields
+
+FROM sample_data* METADATA _index
+| WHERE TO_LONG(@timestamp) < 1698068014937 AND TO_STRING(client_ip) LIKE "172.21.2.16?"
+| KEEP _index, event_duration, message
+| SORT _index ASC, message ASC
+;
+
+_index:keyword       |  event_duration:long  |  message:keyword
+sample_data          |  3450233              |  Connected to 10.1.0.3
+sample_data_str      |  3450233              |  Connected to 10.1.0.3
+sample_data_ts_long  |  3450233              |  Connected to 10.1.0.3
+;
+
+multiIndexWhereIpStringLikeTsLongStats
+required_capability: union_types
+
+FROM sample_data*
+| WHERE TO_LONG(@timestamp) < 1698068014937 AND TO_STRING(client_ip) LIKE "172.21.2.16?"
```

Comment:
If you have a chance, please add a few tests where you convert the same field multiple times to different types, eg.

```
| EVAL x = to_string(@timestamp), y = to_datetime(@timestamp), z= to_long(@timestamp)
```

I tested it and it works fine btw, but it's worth having the coverage ;-)

Reply:
And I think conversions to incompatible types (eg. `to_ip(date)`) have some validation problems. Nothing critical, but it's worth checking.

Probably in these cases we can just return null (we should probably, at least when one of the types is compatible with the conversion), but we could also fail the validation phase with a meaningful error message, up to you.

Reply:
One more thing that just came to my mind:

with multiple indexes with different `@timestamp` fields:

```
from test* | drop @timestamp
```

works just fine

```
from test* | eval x = to_datetime(@timestamp), y = to_string(@timestamp) | drop x, y, @timestamp
```

returns a verification exception:

```
{
    "error": {
        "root_cause": [
            {
                "type": "verification_exception",
                "reason": "Found 2 problems\nline 1:1: Cannot use field [@timestamp] due to ambiguities being mapped as [3] incompatible types: [datetime] in [test2], [keyword] in [test3, test4], [long] in [test1]\nline 1:87: Reference [@timestamp] is ambiguous (to disambiguate use quotes or qualifiers); matches any of [line 1:1 [@timestamp], line 1:35 [@timestamp], line 1:62 [@timestamp]]"
            }
        ],
        "type": "verification_exception",
        "reason": "Found 2 problems\nline 1:1: Cannot use field [@timestamp] due to ambiguities being mapped as [3] incompatible types: [datetime] in [test2], [keyword] in [test3, test4], [long] in [test1]\nline 1:87: Reference [@timestamp] is ambiguous (to disambiguate use quotes or qualifiers); matches any of [line 1:1 [@timestamp], line 1:35 [@timestamp], line 1:62 [@timestamp]]"
    },
    "status": 400
}
```

I think it's due to the fact that now you have multiple `@timestamp` fields.

Also, 

```
from test* | eval x = to_datetime(@timestamp), y = to_string(@timestamp)
```

actually returns multiple `@timestamp` fields (with different types), including the original `unsupported`. I don't think it's what we want... we should at least implicitly them after the renaming, otherwise basic operations like a DROP will become difficult.

Reply:
> so we generate the correct error at planning (validation) time.

I wonder if it makes sense to just short-circuit the wrong conversions, returning null. It would allow to manage field name collisions where a single conversion cannot be applied (eg. when the dataset contains userful data + garbage)

Reply:
Fixing the wrong conversion function was easy enough. The problem was just that when resolving whether a multi-type field mapping was satisfied by an existing convert function, I never checked if the convert function actually supported those types. I just needed to add that check. Then the existing code paths for invalid mapped fields would kick in as normal.

Reply:
But the third issue is more subtle. The idea was to generate a new field with the same name as the old one, and the old one would satisfy use cases where we expect to see `unsupported` and `null`, while the new one would satisfy the use cases where we expect the type to by converted. This seems to have worked great for a wide number of cases, but fails on the `DROP field` command, which I had not tested. It seems like we need to update `DROP` to handle this case. I don't want to remove the new field, because it still covers the case where we do not DROP, and we expect to see the extra field.

Reply:
I think we'll have to hide the newly generated fields anyway, even if we don't have a DROP in the query.

Consider the following:

Reply:
I think the new fields have to be dropped anyway, even without a DROP

Consider the following:

```
PUT index1
{
    "mappings": {
        "properties": {
            "@timestamp": {
                "type": "long"
            },
            "name": {
                "type": "keyword"
            }
        }
    }
}

PUT index2
{
    "mappings": {
        "properties": {
            "@timestamp": {
                "type": "date"
            },
            "name": {
                "type": "keyword"
            }
        }
    }
}
```

```
from test* | eval x = to_datetime(@timestamp)
```

```
{
    "columns": [
        {
            "name": "@timestamp",
            "type": "unsupported"
        },
        {
            "name": "name",
            "type": "keyword"
        },
        {
            "name": "@timestamp",
            "type": "date"
        },
        {
            "name": "x",
            "type": "date"
        }
    ],
    "values": []
}
```

We have two `@timestamp` fields now, that unexpected

Reply:
OK, I've written a few more tests, and indeed if we don't explicitly type `KEEP` we do end up with multiple columns with the same name, as you suspected. So likely the fix to that will also fix the `DROP` error. All my existing tests used `KEEP` to ensure predictable column ordering, but it turns out they had the side effect of hiding this issue.

Reply:
> I don't want to remove the new field, because it still covers the case where we do not DROP, and we expect to see the extra field.

As a user, I expect to only see the value of `x` in this case; the extra `@timestamp` field is an implementation detail that should not appear in the result IMHO

Reply:
It is fixed in the latest commits, including tests that assert the new fields are not in the final output.